db/tags/utils.cc

```diff
@@ -53,4 +55,32 @@ future<> update_tags(service::migration_manager& mm, schema_ptr schema, std::map
     });
 }
 
+future<> modify_tags(service::migration_manager& mm, sstring ks, sstring cf,
+                     std::function<void(std::map<sstring, sstring>&)> modify) {
+    co_await mm.container().invoke_on(0, [ks = std::move(ks), cf = std::move(cf), modify = std::move(modify)] (service::migration_manager& mm) -> future<> {
+        // FIXME: the following needs to be in a loop. If mm.announce() below
+        // fails, we need to retry the whole thing.
+        auto group0_guard = co_await mm.start_group0_operation();
+        // After getting the schema-modification lock, we need to read the
+        // table's *current* schema - it might have changed before we got
+        // the lock, by some concurrent modification. If the table is gone,
+        // this will throw no_such_column_family.
+        schema_ptr s = mm.get_storage_proxy().data_dictionary().find_schema(ks, cf);
+        const std::map<sstring, sstring>* tags_ptr = get_tags_of_table(s);
+        std::map<sstring, sstring> tags;
+        if (tags_ptr) {
+            // tags_ptr is a constant pointer to schema data. To allow func()
+            // to modify the tags, we must make a copy.
+            tags = *tags_ptr;
+        }
+        modify(tags);
+        schema_builder builder(s);
+        builder.add_extension(tags_extension::NAME, ::make_shared<tags_extension>(tags));
+
+        auto m = co_await mm.prepare_column_family_update_announcement(builder.build(), false, std::vector<view_ptr>(), group0_guard.write_timestamp());
+
+        co_await mm.announce(std::move(m), std::move(group0_guard));
```

Comment:
I see a lot of usages of mm.announce together with co_return, I started to wonder, does it make any difference. Just double checking, if  exceptional future is awaited (co_await) it will throw, right?

Reply:
Yes, mm.announce returns an exceptional future, co_await would throw a real exception. We need to handle this exception by the way, but this is an issue I want to handle later (https://github.com/scylladb/scylladb/issues/13152).

Since migration_manager::announce returns a `future<>` I don't understand how `co_return co_await mm.announce` can differ from `co_await mm.announce`, but tell me if you think I'm wrong there.

Reply:
so it's two of us, as I also couldn't figure the difference